feat: Cupertino sheet implement upward stretch on full sheet

masal9pse opened the pull request:
implimented(https://github.com/flutter/flutter/issues/161686)

This PR implements a subtle stretch effect when the user drags the sheet upward.  
It achieves this by dynamically adjusting the top padding while the sheet is fully expanded.

https://github.com/user-attachments/assets/ee98ab82-bc84-40b5-839f-82ae6de59e36

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

masal9pse committed:
feat: implement upward stretch on full sheet; code needs cleanup

masal9pse committed:
add unit test & addressing nits(change arg name etc...)

masal9pse committed:
Merge branch 'master' of https://github.com/flutter/flutter into feature/add-drag-strech-when-max-size

masal9pse committed:
Addressing nits

masal9pse committed:
Merge branch 'master' into feature/add-drag-strech-when-max-size

masal9pse committed:
Merge branch 'master' of https://github.com/flutter/flutter into feature/add-drag-strech-when-max-size

masal9pse committed:
Merge branch 'feature/add-drag-strech-when-max-size' of https://github.com/masal9pse/flutter into feature/add-drag-strech-when-max-size

dkwingsmt commented:
Gentle nudge @MitchellGoodwin (from triage)

masal9pse committed:
Merge branch 'master' of https://github.com/flutter/flutter into feature/add-drag-strech-when-max-size

masal9pse committed:
Add nits

masal9pse committed:
Merge branch 'master' into feature/add-drag-strech-when-max-size

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -735,31 +784,46 @@ class _CupertinoDownGestureDetectorState<T> extends State<_CupertinoDownGestureD
+    // reverse the stretch to return to the normal max height.
+    if (upController.value > 0) {
+      upController.reverse();
+    }
```
Theoretically if the `upController` needs to snap back, then the logic below shouldn't need to happen, right? Can we do a `return;` after this or an `else` to cover the rest of this function?

masal9pse replied:
fix this commit

https://github.com/flutter/flutter/pull/168547/commits/7822c8357f54371d66778e83d7f70f05aaccfbdb

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -735,31 +784,46 @@ class _CupertinoDownGestureDetectorState<T> extends State<_CupertinoDownGestureD
+    // If the sheet is in a stretched state (dragged upward beyond max size),
+    // reverse the stretch to return to the normal max height.
+    if (upController.value > 0) {
+      upController.reverse();
```
Rather than `reverse` it would probably look smoother with an `animateBack` with a non-linear curve. Maybe `easeOut` like the dropped sheet transition.

masal9pse replied:
I see, it is definitely not linear.
As you pointed out, I have modified it to specify an easeOut for animateBack.

https://github.com/flutter/flutter/pull/168547/commits/7131376356c82684f6fa79ed2f5f6c8dcdd24425

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -391,11 +405,20 @@ class _CupertinoSheetTransitionState extends State<CupertinoSheetTransition> {
       reverseCurve: Curves.easeInToLinear,
       parent: widget.secondaryRouteAnimation,
     );
+    _upDragController = AnimationController(
```
Nit: I find the `upDragController`, `downDragController` not immediately obvious. Would something like `stretchDragController` and `popDragController` be more clear?

masal9pse replied:
I see, it is definitely not linear.
As you pointed out, I have modified it to specify an easeOut for animateBack.

https://github.com/flutter/flutter/pull/168547/commits/7131376356c82684f6fa79ed2f5f6c8dcdd24425

MitchellGoodwin reviewed (commented):
Thank you for the PR, this looks great! Left some comments.

masal9pse committed:
Use animateBack with easeOut curve for smoother sheet transition

masal9pse committed:
Change _kMaxTopGapRatio to _kStretchedTopGapRatio.

masal9pse committed:
fix comment

masal9pse committed:
change upDrag to stretchDrag

masal9pse committed:
change downController to popDragController

masal9pse committed:
Add early return when upController snaps back to avoid unnecessary logic

masal9pse committed:
Merge branch 'feature/add-drag-strech-when-max-size' of https://github.com/masal9pse/flutter into feature/add-drag-strech-when-max-size

masal9pse committed:
fix format

masal9pse committed:
Merge remote-tracking branch 'upstream/master' into feature/add-drag-strech-when-max-size

masal9pse commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -29,6 +29,12 @@ const double _kRoundedDeviceCornersThreshold = 20.0;
+// The minimum distance (i.e., maximum upward stretch) from the top of the sheet
+// to the top of the screen, as a ratio of total screen height. This value represents
+// how far the sheet can be temporarily pulled upward before snapping back.
+// Determined through visual tuning to feel natural on iOS 18.0 simulators.
```
fix this commit.

https://github.com/flutter/flutter/pull/168547/commits/156752a2c354e8cd7d2a878d921ed5a759d31191

masal9pse commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -29,6 +29,12 @@ const double _kRoundedDeviceCornersThreshold = 20.0;
+// to the top of the screen, as a ratio of total screen height. This value represents
+// how far the sheet can be temporarily pulled upward before snapping back.
+// Determined through visual tuning to feel natural on iOS 18.0 simulators.
+const double _kMaxTopGapRatio = 0.072;
```
fix this commit.

https://github.com/flutter/flutter/pull/168547/commits/3f80eadf353848addbf8623428ef8b0da538b3db

masal9pse replied:
fix this commit

https://github.com/flutter/flutter/pull/168547/commits/3f80eadf353848addbf8623428ef8b0da538b3db

masal9pse reviewed (commented):
Thanks for pointing that out!
We have addressed all of the points you raised, so please check back when you have a chance!

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -29,6 +29,13 @@ const double _kRoundedDeviceCornersThreshold = 20.0;
+// The minimum distance (i.e., maximum upward stretch) from the top of the sheet
+// to the top of the screen, as a ratio of total screen height. This value represents
+// how far the sheet can be temporarily pulled upward before snapping back.
+// Determined through visual tuning to feel natural on <device name, i.e iPhone 16 Pro>
```
I meant this suggestion as a placeholder. Which device did you simulate when testing this?

masal9pse replied:
Thank you for reviviewing,I tested iPhone16 Pro.

MitchellGoodwin reviewed (approved):
LGTM, just have one nit to figure out for a comment, then I'll ask around for a second review.

masal9pse committed:
Merge branch 'master' of https://github.com/flutter/flutter into feature/add-drag-strech-when-max-size

masal9pse committed:
update commit

masal9pse committed:
fix format

dkwingsmt reviewed (commented):
Looks good except for one comment

masal9pse committed:
Fix missing navigator.didStopUserGesture() in CupertinoSheet stretch handling

Ensures proper gesture state management when sheet is dragged upward beyond max height.

masal9pse committed:
Add null check for _stretchDragController and update it on dependency changes

masal9pse committed:
Merge branch 'master' into feature/add-drag-strech-when-max-size

masal9pse committed:
Merge branch 'master' into feature/add-drag-strech-when-max-size

dkwingsmt reviewed (approved):
LGTM, thank you!

dkwingsmt committed:
Merge branch 'master' into feature/add-drag-strech-when-max-size

dkwingsmt committed:
Merge branch 'master' into feature/add-drag-strech-when-max-size

masal9pse committed:
remove unnecessary diff

masal9pse committed:
Merge branch 'feature/add-drag-strech-when-max-size' of https://github.com/masal9pse/flutter into feature/add-drag-strech-when-max-size
